libbeat/registry/id.go

```diff
@@ -0,0 +1,54 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package registry
+
+import (
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/oklog/ulid"
+)
+
+type idGen rand.Rand
+
+var idPool = sync.Pool{
+	New: func() interface{} {
+		seed := time.Now().UnixNano()
+		rng := rand.New(rand.NewSource(seed))
+		return (*idGen)(rng)
+	},
+}
+
+func newIDGen() *idGen {
+	return idPool.Get().(*idGen)
+}
+
+func (g *idGen) close() {
+	idPool.Put(g)
+}
+
+func (g *idGen) Make() Key {
+	ts := uint64(time.Now().Unix())
+	id := ulid.MustNew(ts, (*rand.Rand)(g))
+	k, err := id.MarshalText()
+	if err != nil {
+		panic(err)
+	}
+	return k
+}
```

Comment:
I had to look a bit into **ulid**, other than the bug that was found in the uuid package. What is the main reason behind using this library it is sorting on id? I am just asking because we introduce a new dependency.

Reply:
ULID is not much better or worse in comparison to UUID, indeed. Some considerations when using ULID vs. UUID:

- UUIDv1 is more similar to ULID, but reserves some bits for the mac address. As ids in the registry file are not shared, the Node-ID is not much of interest. With UUIDv1 we've got 48bit for the random part.
- One advantage of UUIDv1 is the clock-sequence (assuming you are aware of clock changes).
- UUIDv4 is not time-stamped -> reserves 121 (or 122, it depends) bits for the random value. Yet, these 121 are over the complete lifetime of the application
- ULID reserves 48bits for the timestamp in millis + 80 bits for random value.
- ULID is not really sortable, but groups values into quite small 'spans'. This highly reduces chance of duplicates
- ULID is actually vulnerable to timezone changes (summer time vs. winter time), as timestamps will be duplicated, increasing the chance of key duplicates.

Main reason: This implementation configures ULID to be driven by a pseudo rng, whereas the UUID we use requires a cryptographic rng, collecting entropy from the system. Not having enough entropy is the root cause of the bug in the UUID lib -> interface does not provide any more bytes. Using a crypto based rng, we will either have to live with bad/duplicates values due to missing entropy or wait even longer, until we can draw enough samples. Whereas a prng just produceds new samples whenever you need them. For comparison check [these benchmarks](https://github.com/oklog/ulid#benchmarks).

This reminds me, I should check if the key doesn't exist for sure before accepting the generated key. Will update the PR to do so.

In general, the registry store is not supposed to act as a 'heavy' database with thousands of updates per seconds. Mostly as a very small key-value store for a few use-cases we have. I don't really worry about duplicate keys.

Reply:
@urso Sound fine to make, you are right speed shouldn't be an issue with this code.